framework/src/main/java/org/checkerframework/framework/qual/DefaultFor.java

```diff
@@ -7,21 +7,33 @@
 import java.lang.annotation.Target;
 
 /**
- * Applied to the declaration of a type qualifier specifies that the given annotation should be the
- * default for a particular location.
+ * A meta-annotation applied to the declaration of a type qualifier specifies that the given
+ * annotation should be the default for
  *
- * <p>TODO: Document use relative to the other annotations. This qualifier is for type system
- * developers, not end-users.
+ * <ul>
+ *   <li>a particular location.
+ *   <li>a use of a particular type.
+ *   <li>a use of a particular kind of type.
+ * </ul>
  *
  * @see TypeUseLocation
  * @see DefaultQualifier
  * @see DefaultQualifierInHierarchy
- * @see ImplicitFor
  */
 @Documented
 @Retention(RetentionPolicy.RUNTIME)
 @Target(ElementType.ANNOTATION_TYPE)
 public @interface DefaultFor {
     /** @return the locations to which the annotation should be applied */
-    TypeUseLocation[] value();
+    TypeUseLocation[] value() default {};
+
+    /** @return {@link TypeKind}s of types for which an annotation should be implicitly added */
+    TypeKind[] typeKinds() default {};
+
+    /**
+     * @return {@link Class}es for which an annotation should be applied. For example, if
+     *     {@code @MyAnno} is meta-annotated with {@code @DefaultFor(classes=String.class)}, then
+     *     every occurrence of {@code String} is actually {@code @MyAnno String}.
+     */
+    Class<?>[] classes() default {};
```

Comment:
I agree that `typeNames` was confusing, given that it isn't a string attribute. `classes` describes the type of the attribute, but isn't very clear about the intention.
Maybe just `types`?

Reply:
I like `types`.  Here's an example annotation with all three attributes:
```
@DefaultFor(
        value = TypeUseLocation.RETURN, 
        typeKinds = TypeKind.INT,
        types = String.class
)
```
Should we change `value` to `locations`?
```
@DefaultFor(
        locations = TypeUseLocation.RETURN, 
        typeKinds = TypeKind.INT,
        types = String.class
)
```
I kept it as value because most uses of `@DefaultFor` only specify locations; if we change `value` to `locations` it makes those uses more verbose.  (Though it also makes the specification clearer.)

Here's the current annotation for `Initialized`:
```
@DefaultFor({
    TypeUseLocation.IMPLICIT_UPPER_BOUND,
    TypeUseLocation.IMPLICIT_LOWER_BOUND,
    TypeUseLocation.EXCEPTION_PARAMETER
})
```
```
@DefaultFor(
        locations = {
            TypeUseLocation.IMPLICIT_UPPER_BOUND,
            TypeUseLocation.IMPLICIT_LOWER_BOUND,
            TypeUseLocation.EXCEPTION_PARAMETER
        })
```

@mernst, do you have any opinion on the names?

Reply:
@smillst I agree with your preference for `types`.
I am ambivalent about `values` versus `locations`; either is fine with me.

Reply:
I'm going to leave it as `values`.

Reply:
If most locations only use the locations, I'm fine with continuing to use `value`. When there are multiple attributes, using `locations` would be much nicer.